### PR TITLE
Add some prometheus metrics for proxy

### PR DESCRIPTION
**Pull request by xiaocai2333**
Signed-off-by: cai.zhang <cai.zhang@zilliz.com>
issue: #15592

1. Add Prometheus metrics for Proxy

### CODE REVIEW

**Comment by xiaocai2333** [comment]
rerun ut

File: internal/metrics/proxy_metrics.go [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+		}, []string{"client_id", "pchan"})
+
+	// ProxySearchCount record the number of times search succeeded or failed.
+	ProxySearchCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "search_counter",
+			Help:      "The number of times search succeeded or failed",
+		}, []string{"client_id", "status"})
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:49]: status rename to counter

File: internal/metrics/proxy_metrics.go [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+
+	// ProxyQueryCount record The number of times query succeeded or failed.
+	ProxyQueryCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "query_counter",
+			Help:      "The number of times query succeeded or failed",
+		}, []string{"client_id", "status"})
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:58]: rename status to counter

File: internal/metrics/proxy_metrics.go (hunk 1 above)
Comment (czs007) [inline internal/metrics/proxy_metrics.go:49]: Would it be better to keep a metric and then use a type to differentiate search/query?

File: internal/metrics/proxy_metrics.go [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+		}, []string{"client_id"})
+
+	// ProxySearchLatency record the latency of search successfully.
+	ProxySearchLatency = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "search_latency",
+			Help:      "The latency of search successfully",
+		}, []string{"client_id", "type"})
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:113]: Please specify Bucket explicitly, and the unit is milliseconds.

File: internal/metrics/proxy_metrics.go [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+
+	// ProxySendMessageTime record the time that the proxy sent the search request to the message stream.
+	ProxySendMessageTime = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "send_search_msg_time",
+			Help:      "The time that the proxy sent the search request to the message stream",
+		}, []string{"client_id", "type"})
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:122]: Please use type to distinguish search/query.

File: internal/metrics/proxy_metrics.go [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+
+	// ProxyWaitForSearchResultTime record the time that the proxy waits for the search result.
+	ProxyWaitForSearchResultTime = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "wait_for_search_result_time",
+			Help:      "The time that the proxy waits for the search result",
+		}, []string{"client_id", "type"})
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:131]: Please use type to distinguish search/query.

File: internal/metrics/proxy_metrics.go [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+
+	// ProxyReduceSearchResultTime record the time that the proxy reduces search result.
+	ProxyReduceSearchResultTime = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "reduce_search_result_time",
+			Help:      "The time that the proxy reduces search result",
+		}, []string{"client_id", "type"})
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:140]: Please use type to distinguish search/query.

File: internal/metrics/proxy_metrics.go [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+
+	// ProxyDecodeSearchResultTime record the time that the proxy decodes the search result.
+	ProxyDecodeSearchResultTime = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "decode_search_result_time",
+			Help:      "The time that the proxy decodes the search result",
+		}, []string{"client_id", "type"})
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:149]: Please use type to distinguish search/query.

File: internal/metrics/proxy_metrics.go (hunk 2 above)
Comment (czs007) [inline internal/metrics/proxy_metrics.go:58]: add a label named collection_id

File: internal/metrics/proxy_metrics.go [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+		}, []string{"client_id", "status"})
+
+	// ProxyInsertCount record the number of times insert succeeded or failed.
+	ProxyInsertCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "insert_counter",
+			Help:      "The number of times insert succeeded or failed",
+		}, []string{"client_id", "status"})
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:67]: add a label named collection_id

File: internal/metrics/proxy_metrics.go [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+
+	// ProxySearchVectors record the number of vectors search successfully.
+	ProxySearchVectors = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "search_vectors",
+			Help:      "The number of vectors search successfully",
+		}, []string{"client_id"})
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:76]: add a label named collection_id

File: internal/metrics/proxy_metrics.go (hunk 1 above)
Comment (czs007) [reply internal/metrics/proxy_metrics.go:49]: add a label named collection_id

File: internal/metrics/proxy_metrics.go [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+		}, []string{"client_id", "type"})
+
+	// ProxyInsertColumnToRowTime record the time that column to row for inserting in Proxy.
+	ProxyInsertColumnToRowTime = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "insert_column_to_row_time",
+			Help:      "The time that column to row for inserting in Proxy",
+		}, []string{"client_id", "type"})
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:185]: delete type?
add a label named collection_id

File: internal/metrics/proxy_metrics.go [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+		}, []string{"client_id"})
+
+	// ProxyQueryVectors record the number of vectors that query successfully.
+	ProxyQueryVectors = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "query_vectors",
+			Help:      "The number of vectors that query successfully",
+		}, []string{"client_id"})
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:85]: add a label named collection_id
Would it be better to keep a metric and then use a type to differentiate search/query?

File: internal/metrics/proxy_metrics.go [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+
+	// ProxyInsertVectors record the number of vectors insert successfully.
+	ProxyInsertVectors = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "insert_vectors",
+			Help:      "The number of vectors insert successfully",
+		}, []string{"client_id"})
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:94]: add a label named collection_id

File: internal/metrics/proxy_metrics.go (hunk 3 above)
Comment (czs007) [reply internal/metrics/proxy_metrics.go:113]: Please use type to distinguish search/query.

File: internal/metrics/proxy_metrics.go (hunk 4 above)
Comment (czs007) [reply internal/metrics/proxy_metrics.go:122]: Please specify bucket explicitly, and the unit is milliseconds.

File: internal/metrics/proxy_metrics.go (hunk 5 above)
Comment (czs007) [reply internal/metrics/proxy_metrics.go:131]: Please specify bucket explicitly, and the unit is milliseconds.

File: internal/metrics/proxy_metrics.go (hunk 6 above)
Comment (czs007) [reply internal/metrics/proxy_metrics.go:140]: Please specify bucket explicitly, and the unit is milliseconds.

File: internal/metrics/proxy_metrics.go (hunk 4 above)
Comment (czs007) [reply internal/metrics/proxy_metrics.go:122]: add a label named collection_id

File: internal/metrics/proxy_metrics.go [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+	ProxyDecodeSearchResultTime = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "decode_search_result_time",
+			Help:      "The time that the proxy decodes the search result",
+		}, []string{"client_id", "type"})
+
+	// ProxyMsgStreamObjectsPerPChaPerCollection record the number of MsgStream objects per PChannel on each collection on Proxy.
+	ProxyMsgStreamObjectsPerPChaPerCollection = prometheus.NewGaugeVec(
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:152]: name too long

File: internal/metrics/proxy_metrics.go (hunk 6 above)
Comment (czs007) [reply internal/metrics/proxy_metrics.go:140]: add a label named collection_id

File: internal/metrics/proxy_metrics.go (hunk 7 above)
Comment (czs007) [reply internal/metrics/proxy_metrics.go:149]: add a label named collection_id

File: internal/metrics/proxy_metrics.go [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+		prometheus.GaugeOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "MsgStream_objects_per_PChan_per_Collection",
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:156]: name too long.
prometheus will identify this metric as 
<namespace>_<subsystem>_<name>_<type>

File: internal/metrics/proxy_metrics.go [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+			Help:      "The number of MsgStream objects per PChannel on each collection on Proxy",
+		}, []string{"client_id", "Collection"})
+
+	// ProxyMsgStreamObjectsForSearchPerCollection record the number of MsgStream objects for search per collection.
+	ProxyMsgStreamObjectsForSearchPerCollection = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "MsgStream_objects_for_search_per_collection",
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:165]: name too long

File: internal/metrics/proxy_metrics.go [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+		}, []string{"client_id", "collection", "type"})
+
+	// ProxyInsertLatency record the latency that insert successfully.
+	ProxyInsertLatency = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "insert_latency",
+			Help:      "The latency that insert successfully.",
+		}, []string{"client_id", "type"})
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:176]: add a label named collection_id

File: internal/metrics/proxy_metrics.go [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+
+	// ProxyInsertColumnToRowTime record the time that column to row for inserting in Proxy.
+	ProxyInsertColumnToRowTime = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "insert_column_to_row_time",
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:183]: name too long.
add a label named collection_id

File: internal/metrics/proxy_metrics.go [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+			Help:      "The time that column to row for inserting in Proxy",
+		}, []string{"client_id", "type"})
+
+	// ProxySendInsertRequestToMsgStreamTime record the time that Proxy send insert request to MsgStream.
+	ProxySendInsertRequestToMsgStreamTime = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "send_insert_request_to_MsgStream_time",
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:192]: name too long
add a label named collection_id

File: internal/metrics/proxy_metrics.go [hunk 19]
```diff
@@ -0,0 +1,296 @@
...
+	ProxyCacheHitCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "cache_hits",
+			Help:      "Proxy cache hits",
+		}, []string{"client_id", "cacheType", "hitType"})
+
+	// ProxyUpdateCacheTime record the time that proxy update cache when cache miss.
+	ProxyUpdateCacheTime = prometheus.NewHistogramVec(
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:206]: rename to ProxyUpdateCacheLantecy

File: internal/metrics/proxy_metrics.go [hunk 20]
```diff
@@ -0,0 +1,296 @@
...
+	// ProxyCacheHitCounter record the number of Proxy cache hits or miss.
+	ProxyCacheHitCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "cache_hits",
+			Help:      "Proxy cache hits",
+		}, []string{"client_id", "cacheType", "hitType"})
+
+	// ProxyUpdateCacheTime record the time that proxy update cache when cache miss.
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:205]: record the lantency that proxy update cache when case miss

File: internal/metrics/proxy_metrics.go [hunk 21]
```diff
@@ -0,0 +1,296 @@
...
+			Name:      "sync_time_tick",
+			Help:      "Proxy synchronization timestamp statistics, differentiated by Channel",
+		}, []string{"client_id", "channel"})
+
+	// ProxyApplyPrimaryKeyLatency record the latency that apply primary key.
+	ProxyApplyPrimaryKeyLatency = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "apply_primary_key_latency",
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:228]: name too long

File: internal/metrics/proxy_metrics.go [hunk 22]
```diff
@@ -0,0 +1,296 @@
...
+	ProxyFunctionCallCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "function_call_counter",
+			Help:      "Success/failure count for non-insert/query/search",
+		}, []string{"client_id", "function", "status"})
+
+	// ProxyFunctionLatency record the latency that for non-insert/query/search.
+	ProxyFunctionLatency = prometheus.NewHistogramVec(
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:251]: rename to Proxy ReqLantency

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by xiaocai2333** [comment]
rerun ut

File: internal/metrics/proxy_metrics.go [hunk 23]
```diff
@@ -0,0 +1,290 @@
...
+		}, []string{"node_id"})
+
+	// ProxySearchLatency record the latency of search successfully.
+	ProxySearchLatency = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "search_latency",
+			Help:      "The latency of search successfully",
+			Buckets:   []float64{10, 50, 100, 500, 1000}, // unit: ms
```
Comment (LoveEachDay) [inline internal/metrics/proxy_metrics.go:87]: Is there any reason why we choose 10, 50, 100, 500, 1000? Maybe we should record more bucket. setcd server metrics always record 14 buckets.

File: internal/metrics/proxy_metrics.go [hunk 24]
```diff
@@ -0,0 +1,290 @@
...
+			Help:      "The time that the proxy decodes the search result",
+			Buckets:   []float64{5, 10, 20, 50}, // unit: ms
+		}, []string{"node_id", "collection", "type"})
+
+	// ProxyMsgStreamObjectsForPChan record the number of MsgStream objects per PChannel on each collection on Proxy.
+	ProxyMsgStreamObjectsForPChan = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.ProxyRole,
+			Name:      "MsgStream_objects_for_PChan",
```
Comment (LoveEachDay) [inline internal/metrics/proxy_metrics.go:135]: we'd keep the name all in lowercase.

File: internal/metrics/proxy_metrics.go (hunk 23 above)
Comment (xiaocai2333) [reply internal/metrics/proxy_metrics.go:87]: That sounds great, I will add more buckets for Histogram.

File: internal/metrics/proxy_metrics.go [hunk 25]
```diff
@@ -0,0 +1,296 @@
...
+
+package metrics
+
+import (
+	"github.com/milvus-io/milvus/internal/util/typeutil"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	coarseGrainedBuckets = []float64{1, 5, 10, 20, 50, 100, 200, 500, 1000} // unit: ms
```
Comment (czs007) [inline internal/metrics/proxy_metrics.go:25]: may need to add 2000, 5000, 100000

File: internal/proxy/task.go [hunk 26]
```diff
@@ -1341,6 +1348,8 @@ type searchTask struct {
 	chMgr          channelsMgr
 	qc             types.QueryCoord
 	collectionName string
+
+	sendMsgTime time.Time
```
Comment (czs007) [inline internal/proxy/task.go:1352]: use timerecord instead.

**Comment by xiaocai2333** [comment]
rerun ut

**Pull request merged**